<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

In [1]:
%matplotlib inline
%load_ext autoreload
%autoreload 4
%autosave 120

Autosaving every 120 seconds


In [2]:
from fastai.io import *
from fastai.structured import *
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from pandas_summary import DataFrameSummary
from IPython.display import display
from sklearn import metrics
import feather
import lightgbm as lgb
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import mean_squared_error
import warnings

In [40]:
train_clean = pd.read_csv('train_clean.csv')
test_clean = pd.read_csv('test_clean.csv')

In [45]:
train_old = feather.read_dataframe('train_df')
test_old = feather.read_dataframe('test_df')

In [49]:
for df in [train_old, test_old]:
    for f in ['purchase_date_max','purchase_date_min','purchase_date_max_old',\
                     'purchase_date_min_old', 'first_active_month']:
        df[f] = df[f].astype(np.int64) * 1e-9

In [41]:
train_clean0 = train_clean.fillna(0)
train_clean0.to_feather('train_clean0')

In [4]:
def display_all(df):
    with pd.option_context("display.max_rows", 150, "display.max_columns", 100): 
        display(df)

In [9]:
display_all(train_clean.head().T)

0                1  \
first_active_month                              2017-06-01       2017-01-01   
card_id                                    C_ID_92a2005557  C_ID_3d0044924f   
feature_3                                                1                0   
target                                           -0.820283         0.392913   
month                                                    6                1   
year                                                  2017             2017   
elapsed_time                                           245              396   
feature_1_1                                              0                0   
feature_1_2                                              0                0   
feature_1_3                                              0                0   
feature_1_4                                              0                1   
feature_1_5                                              1                0   
feature_2_1                                              0                1   
feature_2_2                                              1                0   
feature_2_3                                              0                0   
authorized_transactions_count                          247              339   
authorized_category_1_sum                                0               29   
authorized_category_1_mean                               0        0.0855457   
authorized_category_3_A_mean                             1                0   
authorized_category_2_5.0_mean                   0.0121457                0   
authorized_category_3_C_mean                             0          0.19174   
authorized_subsector_id_nunique                         21               24   
authorized_merchant_id_nunique                          93              141   
authorized_category_2_1.0_mean                    0.987854         0.914454   
authorized_state_id_nunique                              3                3   
authorized_category_2_4.0_mean                           0                0   
authorized_purchase_amount_sum                    -157.397         -208.883   
authorized_purchase_amount_mean                  -0.637235        -0.616175   
authorized_purchase_amount_max                     2.25839           4.6303   
authorized_purchase_amount_min                   -0.739395          -0.7424   
authorized_purchase_amount_std                    0.216518         0.355553   
authorized_installments_sum                              0              501   
authorized_installments_mean                             0          1.47788   
authorized_installments_max                              0               10   
authorized_installments_min                              0               -1   
authorized_installments_std                              0          1.35063   
authorized_merchant_category_id_nunique                 41               57   
authorized_purchase_month_mean                     7.97976          6.14454   
authorized_purchase_month_max                           12               12   
authorized_purchase_month_min                            1                1   
authorized_purchase_month_std                      3.52857          3.85918   
authorized_month_lag_min                                -8              -12   
authorized_month_lag_max                                 0                0   
authorized_category_2_3.0_mean                           0                0   
authorized_category_2_2.0_mean                           0                0   
authorized_purchase_date_ptp                    2.0978e+07      3.37177e+07   
authorized_purchase_date_min                   1.49857e+09      1.48372e+09   
authorized_purchase_date_max                   1.51955e+09      1.51744e+09   
authorized_category_3_B_mean                             0          0.80236   
authorized_city_id_nunique                               7                9   
unauthorized_transactions_co

In [10]:
display_all(test_clean.head().T)

0                1  \
first_active_month                              2017-04-01       2017-01-01   
card_id                                    C_ID_0ab67a22ab  C_ID_130fd0cbdd   
feature_3                                                1                0   
month                                                    4                1   
year                                                  2017             2017   
elapsed_time                                           306              396   
feature_1_1                                              0                0   
feature_1_2                                              0                1   
feature_1_3                                              1                0   
feature_1_4                                              0                0   
feature_1_5                                              0                0   
feature_2_1                                              0                0   
feature_2_2                                              0                0   
feature_2_3                                              1                1   
authorized_transactions_count                           44               77   
authorized_category_1_sum                               10                1   
authorized_category_1_mean                        0.227273         0.012987   
authorized_category_3_A_mean                             0                0   
authorized_category_2_5.0_mean                           0                0   
authorized_category_3_C_mean                      0.204545         0.025974   
authorized_subsector_id_nunique                         12               12   
authorized_merchant_id_nunique                          23               27   
authorized_category_2_1.0_mean                    0.772727                0   
authorized_state_id_nunique                              3                3   
authorized_category_2_4.0_mean                           0         0.974026   
authorized_purchase_amount_sum                    -28.0385         -48.4046   
authorized_purchase_amount_mean                  -0.637239        -0.628632   
authorized_purchase_amount_max                    0.235676         0.318818   
authorized_purchase_amount_min                   -0.743902        -0.731881   
authorized_purchase_amount_std                    0.169498         0.155571   
authorized_installments_sum                             63               82   
authorized_installments_mean                       1.43182          1.06494   
authorized_installments_max                              6                4   
authorized_installments_min                              1                1   
authorized_installments_std                        1.14927         0.408388   
authorized_merchant_category_id_nunique                 15               16   
authorized_purchase_month_mean                     8.52273          3.25974   
authorized_purchase_month_max                           12                5   
authorized_purchase_month_min                            4                1   
authorized_purchase_month_std                      2.57437          1.35144   
authorized_month_lag_min                                -8              -13   
authorized_month_lag_max                                 0                0   
authorized_category_2_3.0_mean                           0         0.012987   
authorized_category_2_2.0_mean                           0                0   
authorized_purchase_date_ptp                   2.31794e+07      3.46679e+07   
authorized_purchase_date_min                   1.49133e+09      1.48432e+09   
authorized_purchase_date_max                   1.51451e+09      1.51899e+09   
authorized_category_3_B_mean                      0.795455         0.974026   
authorized_city_id_nunique                               7                4   
unauthorized_transactions_count                         24                1   
unauthorized_category_1_sum 

In [5]:
def print_score(m, x_trn, y_trn, x_val, y_val):
    res = [rmse(m.predict(x_trn), y_trn), rmse(m.predict(x_val), y_val),
                m.score(x_trn, y_trn), m.score(x_val, y_val)]
    if hasattr(m, 'oob_score_'): res.append(m.oob_score_)
    print(res)

In [6]:
def rmse(preds, target): 
    return np.sqrt(mean_squared_error(preds, target))

In [42]:
rf_args = dict({
    "n_estimators": 20, 
    "min_samples_leaf":3, 
    "max_features": 0.9, 
    "n_jobs":-1, 
    "oob_score":True
})

In [8]:
target = train_clean['target']

In [10]:
train_clean.size

24230040

In [13]:
train_clean.shape

(201917, 120)

In [14]:
train_df = train_clean.copy()

In [15]:
add_datepart(train_df, 'first_active_month')

In [16]:
display_all(train_df.dtypes)

card_id                                       object
feature_3                                      int64
month                                          int64
year                                           int64
elapsed_time                                   int64
feature_1_1                                    int64
feature_1_2                                    int64
feature_1_3                                    int64
feature_1_4                                    int64
feature_1_5                                    int64
feature_2_1                                    int64
feature_2_2                                    int64
feature_2_3                                    int64
authorized_transactions_count                  int64
authorized_category_1_sum                    float64
authorized_category_1_mean                   float64
authorized_category_3_A_mean                 float64
authorized_category_2_5.0_mean               float64
authorized_category_3_C_mean                 f

In [18]:
train_df.head()

card_id  feature_3  month  year  elapsed_time  feature_1_1  \
0  C_ID_92a2005557          1      6  2017           245            0   
1  C_ID_3d0044924f          0      1  2017           396            0   
2  C_ID_d639edf6cd          0      8  2016           549            0   
3  C_ID_186d6a6901          0      9  2017           153            0   
4  C_ID_cdbd2c0db2          0     11  2017            92            1   

   feature_1_2  feature_1_3  feature_1_4  feature_1_5  \
0            0            0            0            1   
1            0            0            1            0   
2            1            0            0            0   
3            0            0            1            0   
4            0            0            0            0   

             ...              first_active_monthDay  \
0            ...                                  1   
1            ...                                  1   
2            ...                                  1   
3            ...                                  1   
4            ...                                  1   

   first_active_monthDayofweek  first_active_monthDayofyear  \
0                            3                          152   
1                            6                            1   
2                            0                          214   
3                            4                          244   
4                            2                          305   

   first_active_monthIs_month_end  first_active_monthIs_month_start  \
0                           False                              True   
1                           False                              True   
2                           False                              True   
3                           False                              True   
4                           False                              True   

   first_active_monthIs_quarter_end  first_active_monthIs_quarter_start  \
0                             False                               False   
1                             False                                True   
2                             False                               False   
3                             False                               False   
4                             False                               False   

   first_active_monthIs_year_end  first_active_monthIs_year_start  \
0                          False                            False   
1                          False                             True   
2                          False                            False   
3                          False                            False   
4                          False                            False   

   first_active_monthElapsed  
0                 1496275200  
1                 1483228800  
2                 1470009600  
3                 1504224000  
4                 1509494400  

[5 rows x 132 columns]

In [19]:
train_df.drop('card_id', axis=1, inplace=True)

In [21]:
train_df2 = train_df.fillna(0)

In [43]:
train_df2.drop('outliers', axis=1, inplace=True)

In [22]:
train_cats(train_df2)

In [51]:
train_old.drop('card_id', axis=1, inplace=True)

In [53]:
df_hot_ext = feather.read_dataframe('df_hot_ext')
df_hot_test_ext = feather.read_dataframe('df_hot_test_ext')

In [55]:
for df in [df_hot_ext, df_hot_test_ext]:
    for f in ['purchase_date_amax','purchase_date_amin','purchase_date_amax_old',\
                     'purchase_date_amin_old']:
        df[f] = df[f].astype(np.int64) * 1e-9

In [62]:
df_hot_ext.drop('is_outlier', axis=1, inplace=True)

In [60]:
DataFrameSummary(df_hot_ext).summary().T

count         mean          std  \
month_diff_mean_old                   201917      11.7556      2.39709   
category_1_sum_old                    201917      6.37936      19.5725   
authorized_flag_mean_old              201917     0.893353     0.114446   
purchase_amount_max                   201917    -0.173108      1.45139   
month_lag_mean_old                    201917     -3.84099      2.12349   
purchase_date_amax_old                201917      1.51585   0.00645884   
category_1_mean_old                   201917      0.10108     0.208051   
purchase_date_amin_old                201917      1.49475    0.0102942   
purchase_amount_min_old               201917    -0.729993    0.0918995   
hist_first_buy                        201917      118.748      224.136   
purchase_date_ptp_old                 201917      243.678      118.292   
purchase_amount_mean_old              201917      2.45943      1115.17   
purchase_amount_max_old               201917      66.6717      13448.5   
purchased_on_weekend_mean_old         201917     0.276675     0.138688   
month_lag_sum_old                     201917     -400.169      610.571   
inverse_avg_transactions_per_day_old  201917      6.07849      6.93816   
days_since_last_transaction           201917       303.57      69.9628   
purchase_date_amax                    201917     0.353809       3.3431   
purchase_amount_std_old               201917      12.6537      3865.24   
installments_mean_old                 201917     0.863678      1.13581   
month_diff_var_old                    201917     0.217643    0.0521406   
purchase_amount_mean                  201917    -0.537137     0.525156   
purchase_amount_min                   201917    -0.667412     0.409663   
month_diff_std_old                    201917     0.459549    0.0803638   
purchase_amount_sum                   201917     -3.62723      4.16273   
month_lag_std_old                     201917      2.40891      1.17382   
days_since_last_transaction_old       201917      363.735      74.7394   
merchant_id_nunique_old               201917      33.7136      30.6869   
merchant_category_id_nunique_old      201917       18.278      11.4705   
new_first_buy                         201917      372.072      270.356   
...                                      ...          ...          ...   
month_diff_mean                       201917      11.6172      2.29955   
authorized_flag_sum_old               201917       81.559      99.2434   
first_active_monthDayofyear           201917      194.482      101.796   
city_id_nunique_old                   201917      4.90163      3.49626   
first_active_monthWeek                201917      30.9446      13.8454   
elapsed_time                          201917      726.979       293.71   
purchased_on_weekend_mean             201917     0.288241     0.271316   
installments_mean                     201917     0.699745      1.08214   
state_id_nunique_old                  201917      2.73086      1.51666   
card_id_total                         201917       96.107      104.283   
first_active_month                    201917  1.48444e+09  2.53766e+07   
category_1_mean                       201917    0.0438083     0.155868   
feature_1                             201917    0.0109302    0.0012591   
installments_sum                      201917      4.23266      7.81939   
month_diff_var                        201917     0.211563     0.121516   
installments_max_old                  201917      4.25316      22.9412   
month_lag_max_old                     201917    -0.166058     0.636972   
merchant_category_id_nunique          201917      5.19377      4.03915   
month_lag_min_old                     201917     -7.87778      3.84331   
month_diff_std                        201917     0.410666     0.207164   
month_lag_sum                         201917      9.67641      9.57903   
installments_std                      201917     0.394333     0.979057   
feature_2                             20191

In [63]:
m = RandomForestRegressor(**rf_args)
m.fit(df_hot_ext, target)
m.score(df_hot_ext, target)
print_score(m, df_hot_ext, target.values, df_hot_ext, target.values)

/home/chandrasekhar/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/ensemble/forest.py:732: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable oob estimates.
  warn("Some inputs do not have OOB scores. "


[2.303248285113751, 2.303248285113751, 0.6421925248673591, 0.6421925248673591, -0.023896062366220994]


- Base model - 0.8348232234871118, 0.8348232234871118, 0.9529936055779922, 0.9529936055779922, 0.8147359464393436
- Base model with outliers removed - 2.299946409218127, 2.299946409218127, 0.6432176762133195, 0.6432176762133195, -0.02364908316679548
- Base model with df_hot_ext - 2.1145840168101624, 2.1145840168101624, 0.6984094099753392, 0.6984094099753393, 0.1985261186716828
- Base model with df_hot_ext ouliers removed - 2.303248285113751, 2.303248285113751, 0.6421925248673591, 0.6421925248673591, -0.023896062366220994

In [34]:
display_all(train_clean.isna().sum().sort_values(ascending=False))

unauthorized_purchase_amount_std             51988
unauthorized_purchase_month_std              51988
unauthorized_installments_std                51988
new_installments_std                         48718
new_purchase_month_std                       48718
new_purchase_amount_std                      48718
unauthorized_month_lag_max                   30959
unauthorized_month_lag_min                   30959
unauthorized_category_2_3.0_mean             30959
unauthorized_installments_sum                30959
unauthorized_category_2_2.0_mean             30959
unauthorized_purchase_month_mean             30959
unauthorized_merchant_category_id_nunique    30959
unauthorized_purchase_date_ptp               30959
unauthorized_installments_min                30959
unauthorized_purchase_date_min               30959
unauthorized_installments_max                30959
unauthorized_installments_mean               30959
unauthorized_purchase_month_min              30959
unauthorized_category_2_4.0_mea

In [35]:
train_clean0 = train_clean.fillna(0)

In [37]:
train_clean0.isna().sum().sort_values(ascending=False)

outliers                                     0
new_city_id_nunique                          0
authorized_installments_mean                 0
authorized_installments_max                  0
authorized_installments_min                  0
authorized_installments_std                  0
authorized_merchant_category_id_nunique      0
authorized_purchase_month_mean               0
authorized_purchase_month_max                0
authorized_purchase_month_min                0
authorized_purchase_month_std                0
authorized_month_lag_min                     0
authorized_month_lag_max                     0
authorized_category_2_3.0_mean               0
authorized_category_2_2.0_mean               0
authorized_purchase_date_ptp                 0
authorized_purchase_date_min                 0
authorized_purchase_date_max                 0
authorized_category_3_B_mean                 0
authorized_city_id_nunique                   0
unauthorized_transactions_count              0
unauthorized_

In [38]:
train_clean0.to_feather('train_clean0')